src/cmd/job/mpir.c

```diff
@@ -74,6 +79,160 @@ static void gen_attach_signal (flux_t *h, flux_jobid_t id)
+    const char *prog = basename (MPIR_executable_path);
+
+    if (signum > 0)
+        log_msg ("MPIR: rank %d: %s: %s", rank, prog, strsignal (signum));
+    else if (exitcode != 0)
+        log_msg ("MPIR: rank %d: %s: Exit %d", rank, prog, exitcode);
+    flux_subprocess_destroy (p);
+}
+
+static flux_cmd_t *mpir_make_tool_cmd (const char *path,
+                                       const char *server_args,
+                                       int server_args_len)
+{
+    flux_cmd_t *cmd = NULL;
+    char *dir = NULL;
+    char **argv;
+    int argc;
+
+    if (mpir_create_argv (path,
+                          server_args,
+                          server_args_len,
+                          &argc,
+                          (const char ***) &argv) < 0)
+        return NULL;
+
+    if (argc == 0
+        || !(cmd = flux_cmd_create (argc, argv, environ))) {
+        log_err ("failed to create command from MPIR_executable_path");
+        return NULL;
+    }
+
+    flux_cmd_unsetenv (cmd, "FLUX_PROXY_REMOTE");
+    if (!(dir = get_current_dir_name ())
+        || flux_cmd_setcwd (cmd, dir) < 0)
+        log_err_exit ("failed to get or set current directory");
+    free (argv);
+    free (dir);
+    return cmd;
+}
+
+static void output_cb (flux_subprocess_t *p, const char *stream)
+{
+    const char *line;
+    int len = 0;
+    const char *prog = basename (MPIR_executable_path);
+    int rank = flux_subprocess_rank (p);
+
+    line = flux_subprocess_read_trimmed_line (p, stream, &len);
+    if (line && len == 0)
+        line = flux_subprocess_read (p, stream, -1, &len);
+    if (len)
+        log_msg ("MPIR: rank %d: %s: %s: %s", rank, prog, stream, line);
+}
+
+static void state_cb (flux_subprocess_t *p, flux_subprocess_state_t state)
+{
+    if (state == FLUX_SUBPROCESS_FAILED) {
+        const char *prog = basename (MPIR_executable_path);
+        int rank = flux_subprocess_rank (p);
+        const char *errmsg = flux_subprocess_fail_error (p);
+        log_msg ("MPIR: rank %d: %s: %s", rank, prog, errmsg);
+    }
+}
+
+static void launch_tool_daemons (flux_t *h,
+                                 const char *tool_path,
+                                 const char *tool_args,
+                                 int tool_args_len,
+                                 struct idset *ranks)
+{
+    unsigned int rank;
+    flux_cmd_t *cmd;
+
+    flux_subprocess_ops_t ops = {
+        .on_completion = completion_cb,
+        .on_stdout = output_cb,
+        .on_stderr = output_cb,
+        .on_state_change = state_cb,
+    };
+
+    if (!(cmd = mpir_make_tool_cmd (tool_path, tool_args, tool_args_len)))
+        return;
+
+    rank = idset_first (ranks);
+    while (rank != IDSET_INVALID_ID) {
+        flux_subprocess_t *p;
+        if (!(p = flux_rexec (h, rank, 0, cmd, &ops)))
+            log_err ("MPIR: failed to launch %s", tool_path);
+        rank = idset_next (ranks, rank);
+    }
+    flux_cmd_destroy (cmd);
+    return;
```

Comment:
Are the subprocess objects leaked here purposefully because there is no way to return an error via the MPIR interface?